### PR TITLE
Fix/367: Create and Update User not populated correctly

### PR DESCRIPTION
**Pull request by NickSaglioni**
Passing the ID token to the backend and handling it for idp_username

### CODE REVIEW

**Comment by ianliuwk1019** [comment]
I feel code can be improved from using the draft branch "feat/359-auth-dependency-injection-partial-work".
As it definite can be setup as a way to get the token without explicitly retrieved the request header for ID token after router is called. But that can be later in tech session for it. Just a suggestion.

File: server/backend/api/app/crud/crud_user_role.py [hunk 1]
```diff
@@ -34,7 +35,7 @@ def create_user_role(
         crud_utils.raise_http_exception(HTTPStatus.BAD_REQUEST, error_msg)
 
     # Determine if user already exists or add a new user.
-    fam_user = crud_user.find_or_create(db, request.user_type_code, request.user_name)
+    fam_user = crud_user.find_or_create(db, request.user_type_code, request.user_name, user_creator)
```
Comment (ianliuwk1019) [inline server/backend/api/app/crud/crud_user_role.py:38]: Same comment for the variable naming, and other places.

File: server/backend/api/app/jwt_validation.py [hunk 2]
```diff
@@ -228,3 +230,29 @@ def authorize_by_app_id(
...
+                    'description':
+                    'Unable to decode ID token.'},
+            headers={"WWW-Authenticate": "Bearer"},
+        )
+
+    if JWT_USERNAME_KEY not in unverified_claims:
+        raise HTTPException(
+            status_code=403,
+            detail={'code': ERROR_USERNAME_REQUIRED,
+                    'description': 'Required custom:idp_username in claim'},
```
Comment (ianliuwk1019) [inline server/backend/api/app/jwt_validation.py:254]: Maybe 'Requires custom:idp_username in claim'

File: server/backend/api/app/jwt_validation.py [hunk 3]
```diff
@@ -228,3 +230,29 @@ def authorize_by_app_id(
...
+            status_code=401,
+            detail={'code': ERROR_TOKEN_DECODE,
+                    'description':
+                    'Unable to decode ID token.'},
+            headers={"WWW-Authenticate": "Bearer"},
+        )
+
+    if JWT_USERNAME_KEY not in unverified_claims:
+        raise HTTPException(
+            status_code=403,
```
Comment (ianliuwk1019) [inline server/backend/api/app/jwt_validation.py:252]: Should this be 401?

**Review by ianliuwk1019 — CHANGES_REQUESTED** [review]

File: frontend/src/services/http/HttpRequestInterceptors.ts [hunk 4]
```diff
@@ -9,13 +9,14 @@ function addAuthHeaderItcpt(config: AxiosRequestConfig) {
     const authToken = authService.state.value.famLoginUser?.authToken
     if (authToken) {
         const authHeader = `Bearer ${authToken.getAccessToken().getJwtToken()}`
-        config.headers? config.headers['Authorization'] = authHeader
-                      : config.headers = {'Authorization': authHeader}
+        const idTokenHeader = authToken.getIdToken().getJwtToken()
+        config.headers? (config.headers['Authorization'] = authHeader, config.headers['id-token'] = idTokenHeader)
+                      : (config.headers = {'Authorization': authHeader}, config.headers = {'id-token': idTokenHeader})
```
[the remark this answers is not in the document]
Comment (NickSaglioni) [reply frontend/src/services/http/HttpRequestInterceptors.ts:14]: Yes, this works.

File: server/backend/api/app/crud/crud_user.py [hunk 5]
```diff
@@ -97,7 +97,7 @@ def delete_user(db: Session, user_id: int):
     return fam_user
 
 
-def find_or_create(db: Session, user_type_code: str, user_name: str):
+def find_or_create(db: Session, user_type_code: str, user_name: str, user_creator: str):
```
[the remark this answers is not in the document]
Comment (NickSaglioni) [reply server/backend/api/app/crud/crud_user.py:100]: Changing to 'requester'

File: server/backend/api/app/crud/crud_user_role.py (hunk 1 above)
Comment (NickSaglioni) [reply server/backend/api/app/crud/crud_user_role.py:38]: Same as above

File: server/backend/api/app/jwt_validation.py (hunk 2 above)
Comment (NickSaglioni) [reply server/backend/api/app/jwt_validation.py:254]: Thanks, correcting it

File: server/backend/api/app/jwt_validation.py (hunk 3 above)
Comment (NickSaglioni) [reply server/backend/api/app/jwt_validation.py]: Thanks! Fixed

**Review by ianliuwk1019 — APPROVED** [review]
Thanks Nick!
Added one more requester in crud_forest_client for "create_user"
Data verified.
Approve this PR.

**Pull request merged**